--- a/source/Makefile.in
+++ b/source/Makefile.in
@@ -476,11 +476,11 @@ bin/smbfilter: $(SMBFILTER_OBJ) bin/.dum
 
 bin/smbwrapper.@SHLIBEXT@: $(PICOBJS)
 	@echo Linking shared library $@
-	@$(LD) @LDSHFLAGS@ -o $@ $(PICOBJS) $(LIBS)
+	@$(LD) @LDSHFLAGS@ -o $@ $(PICOBJS) $(LDFLAGS) $(LIBS)
 
 bin/smbwrapper.32.@SHLIBEXT@: $(PICOBJS32)
 	@echo Linking shared library $@
-	@$(LD) -32 @LDSHFLAGS@ -o $@ $(PICOBJS32) $(LIBS)
+	@$(LD) -32 @LDSHFLAGS@ -o $@ $(PICOBJS32) $(LDFLAGS) $(LIBS)
 
 bin/smbsh: $(SMBSH_OBJ) bin/.dummy
 	@echo Linking $@
@@ -488,7 +488,7 @@ bin/smbsh: $(SMBSH_OBJ) bin/.dummy
 
 bin/libsmb.@SHLIBEXT@: $(LIBSMB_PICOBJS) bin/.dummy
 	@echo Linking shared library $@
-	@$(LD) @LDSHFLAGS@ -o $@ $(LIBSMB_PICOBJS) $(LIBS)
+	@$(LD) @LDSHFLAGS@ -o $@ $(LIBSMB_PICOBJS) $(LDFLAGS) $(LIBS)
 
 install: installbin installman installscripts installcp installswat
 
