--- a/net/ipv4/netfilter/nf_nat_proto_udp.c
+++ b/net/ipv4/netfilter/nf_nat_proto_udp.c
@@ -65,6 +65,10 @@ udp_unique_tuple(struct nf_conntrack_tup
 				min = 600;
 				range_size = 1023 - min + 1;
 			}
+		} else if (ntohs(*portptr) == 5060) {
+			/* Miguel GAIO: we wanted SIP dport closed to 5060 */
+			min = 5061;
+			range_size = 65535 - 5061 + 1;
 		} else {
 			min = 1024;
 			range_size = 65535 - 1024 + 1;
--- a/net/netfilter/nf_conntrack_sip.c
+++ b/net/netfilter/nf_conntrack_sip.c
@@ -336,7 +336,8 @@ int ct_sip_get_info(struct nf_conn *ct,
 
 	while (dptr <= limit) {
 		if ((strncmp(dptr, hnfo->lname, hnfo->lnlen) != 0) &&
-		    (strncmp(dptr, hnfo->sname, hnfo->snlen) != 0)) {
+		    (hnfo->sname == NULL ||
+		     strncmp(dptr, hnfo->sname, hnfo->snlen) != 0)) {
 			dptr++;
 			continue;
 		}
