--- a/net/ipv4/netfilter/ipt_TOS.c
+++ b/net/ipv4/netfilter/ipt_TOS.c
@@ -57,8 +57,8 @@ checkentry(const char *tablename,
 	    && tos != IPTOS_RELIABILITY
 	    && tos != IPTOS_MINCOST
 	    && tos != IPTOS_NORMALSVC) {
-		printk(KERN_WARNING "TOS: bad tos value %#x\n", tos);
-		return 0;
+		printk(KERN_WARNING "TOS: bad tos value %#x, accepting\n", tos); /* Efixo mods: accept unnormalized TOS */
+		return 1;
 	}
 	return 1;
 }
