--- a/networking/udhcp/packet.c
+++ b/networking/udhcp/packet.c
@@ -216,6 +216,9 @@ int FAST_FUNC udhcp_send_raw_packet(stru
 	packet.ip.ihl = sizeof(packet.ip) >> 2;
 	packet.ip.version = IPVERSION;
 	packet.ip.ttl = IPDEFTTL;
+#ifdef VOIP_DSCP
+	packet.ip.tos = (VOIP_DSCP << 2); /* neufbox TOS */
+#endif
 	packet.ip.check = udhcp_checksum(&packet.ip, sizeof(packet.ip));
 
 	udhcp_dump_packet(dhcp_pkt);
@@ -238,6 +241,9 @@ int FAST_FUNC udhcp_send_kernel_packet(s
 {
 	struct sockaddr_in client;
 	int fd;
+#ifdef VOIP_DSCP
+	int tos;
+#endif
 	int result = -1;
 	const char *msg;
 
@@ -248,6 +254,11 @@ int FAST_FUNC udhcp_send_kernel_packet(s
 	}
 	setsockopt_reuseaddr(fd);
 
+#ifdef VOIP_DSCP
+	tos = (VOIP_DSCP << 2); /* neufbox TOS */
+	setsockopt(fd, SOL_IP, IP_TOS, &tos, sizeof(tos));
+#endif
+
 	memset(&client, 0, sizeof(client));
 	client.sin_family = AF_INET;
 	client.sin_port = htons(source_port);
