--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -110,6 +110,7 @@ static void macvlan_broadcast(struct sk_
 	struct net_device *dev;
 	struct sk_buff *nskb;
 	unsigned int i;
+	unsigned is_multicast;
 
 	if (skb->protocol == htons(ETH_P_PAUSE))
 		return;
@@ -118,6 +119,12 @@ static void macvlan_broadcast(struct sk_
 		hlist_for_each_entry_rcu(vlan, n, &port->vlan_hash[i], hlist) {
 			dev = vlan->dev;
 
+			is_multicast = (!!compare_ether_addr(eth->h_dest, dev->broadcast));
+			if (is_multicast && (!(dev->flags & IFF_MULTICAST))) {
+				dev->stats.rx_dropped++;
+				continue;
+			}
+
 			nskb = skb_clone(skb, GFP_ATOMIC);
 			if (nskb == NULL) {
 				dev->stats.rx_errors++;
@@ -130,7 +137,7 @@ static void macvlan_broadcast(struct sk_
 			dev->stats.multicast++;
 
 			nskb->dev = dev;
-			if (!compare_ether_addr(eth->h_dest, dev->broadcast))
+			if (!is_multicast)
 				nskb->pkt_type = PACKET_BROADCAST;
 			else
 				nskb->pkt_type = PACKET_MULTICAST;
