--- a/drivers/leds/leds-pwm.c
+++ b/drivers/leds/leds-pwm.c
@@ -47,6 +47,14 @@ static void led_pwm_set(struct led_class
 	}
 }
 
+static enum led_brightness led_pwm_get(struct led_classdev *led_cdev)
+{
+	struct led_pwm_data *led_dat =
+		container_of(led_cdev, struct led_pwm_data, cdev);
+
+	return pwm_get_config(led_dat->pwm);
+}
+
 static int led_pwm_probe(struct platform_device *pdev)
 {
 	struct led_pwm_platform_data *pdata = pdev->dev.platform_data;
@@ -80,6 +88,7 @@ static int led_pwm_probe(struct platform
 		led_dat->max_brightness = cur_led->max_brightness;
 		led_dat->period = cur_led->pwm_period_ns;
 		led_dat->cdev.brightness_set = led_pwm_set;
+		led_dat->cdev.brightness_get = led_pwm_get;
 		led_dat->cdev.brightness = LED_OFF;
 		led_dat->cdev.flags |= LED_CORE_SUSPENDRESUME;
 
--- a/include/linux/pwm.h
+++ b/include/linux/pwm.h
@@ -19,6 +19,11 @@ void pwm_free(struct pwm_device *pwm);
 int pwm_config(struct pwm_device *pwm, int duty_ns, int period_ns);
 
 /*
+ * pwm_config - get PWM device configuration
+ */
+int pwm_get_config(struct pwm_device *pwm);
+
+/*
  * pwm_enable - start a PWM output toggling
  */
 int pwm_enable(struct pwm_device *pwm);
